docs/style-guide.md

```diff
@@ -85,27 +121,77 @@ There are three areas which may warrant a technically public API that we don't c
+}
+```
+
+Since the `date_style` and `time_style` fields can accept only valid values, and the getter/setter methods
+would not perform any additional operations, those fields can be exposed as public.
+
+In all other situations, where the getter/setter are fallible, perform additional computations or optimizations, or an invariant between the fields has to be achieved, it is advised to use an equivalent getter/setter model:
 
-Unless there is a compelling use case, it seems very sensible to never use public fields in structs. This allows for clean encapsulation with essentially no downside for ICU4X. This post sums things up nicely, which I am summarizing below:
+```rust
+#[derive(Default)]
+struct DateTimeOptions {
+    meta: Option<TinyStr8>,
+}
 
-For a **pub** field `bar` in a **pub** struct `Foo`:
+impl DateTimeOptions {    
+    pub fn get_meta(&self) -> Option<&str>;
+    pub fn set_meta(&mut self, input: &str) -> Result;
+}
 
-* Construction: users can create a `Foo { bar: 42 }`.
-  * This also means it’s a breaking change for you to add or remove any fields (and using a Default trait won't help you here since you still have to rely on every caller using it to avoid adding a new field being a breaking change).
-* Mutation: if the user has some `mut var: Foo` or `&mut Foo`, they can change the value without notice.
-  * While this won't cause any race conditions in Rust, it does prevent a whole class of useful optimizations for types around internal consistency (e.g. you cannot cache otherwise expensive hash-code values).
-* Partial borrow: if a user binds `let var = &foo.bar;` they can still use other fields without any complaint from the borrow checker.
+fn main() {
+    let mut opts = DateTimeOptions::default();
+  
+    // Override
+    opts.set_meta("foo")?;
+}
+```
 
-You can always supply public inlinable getters to access fields, and non-public fields are still directly accessible by the current module and any sub-modules.
+Note: Any change to field visibility constitute a breaking change to the public API.
+Note: Even if the setter is infallible, the getter/setter model is useful when optimized type is used internally while public API exposes a standard type, like in the case of `meta` field in the example above.
 
-One suggested situation in which public fields would be acceptable is for user-facing "bag of options" structs. These would have no inbuilt semantics and no consistency guarantees, so the visual "cleanliness" of bare fields might outweigh the issues above. See [this issue](https://github.com/unicode-org/rust-discuss/issues/15) for more.
+One suggested situation in which public fields would be acceptable is for user-facing "bag of options" structs. These would have no inbuilt semantics and no consistency guarantees, so the superior ergonomics of the public fields can be benefited from. In that case, it is recommended to also implement or derive the `Default` trait. See `Constructor` section below for details.
```

Comment:
If we add a field to an option bag it is a breaking change and we need to iterate our release version. Using `Default` doesn't save us here.

You can find this paragraph in the style guide above: "This also means it’s a breaking change for you to add or remove any fields (and using a Default trait won't help you here since you still have to rely on every caller using it to avoid adding a new field being a breaking change)."

Reply:
I know adding fields to a struct is breaking, but I really don't want to have to bump major version numbers to add options, because that's something we may want to do frequently.  I've been hoping that if we can put a lot of weight behind the `Default` trait, we can avoid this issue.

Did we consider requiring all structs to have one private dummy field to force clients to use `Default`?

Reply:
I don't know of any prior work for that and it seems quirky to try. @Manishearth - what's your take?

I'd prefer to assume that the `Options` struct is public and changing its fields is breaking.

Reply:
You can just tag the options bag as non exhaustive.

Reply:
For example, something like this:

```rust
pub struct DateTimeOptions {
    _dummy: (),
    date_style: Option<DateStyle>,
    time_style: Option<TimeStyle>,
}
```

That's a little annoying, but I think it's much better than making a major version bump to add a field to a struct.  Major version bumps should be the exception, not the rule.

Reply:
@sffc again, there is a non_exhaustive tag you can use that solves this, no need to hack around it.

Reply:
that sounds like a good solution! I'll add it to the style guide.

Reply:
Added the `#[non_exhaustive]` attribute to all `Options` and added a paragraph about its use.

Reply:
`#[non_exhaustive]` is great!  Thanks for the pointer!